PIMbench/cpp-knn/baselines/CPU/knn.cpp

```diff
@@ -0,0 +1,280 @@
+                    if (int(localMinHeaps[i].size()) < k) {
+                        localMinHeaps[i].emplace(dist, j);
+                    } else if (dist < localMinHeaps[i].top().distance) {
+                        localMinHeaps[i].pop();
+                        localMinHeaps[i].emplace(dist, j);
+                    }
+
+                }
+            }
+#pragma omp critical
+            {
+                for (int i = 0; i < numTests; ++i) {
+                    // Tally the labels of the k nearest neighbors
+                    unordered_map<int, int> labelCount;
+                    while (!localMinHeaps[i].empty()) {
+                        int index = localMinHeaps[i].top().index;
+                        int label = dataPoints[index][target];
+                        labelCount[label]++;
+                        localMinHeaps[i].pop();
+                    }
+                    
+                    // Find the label with the highest count
+                    int maxCount = 0;
+                    int bestLabel = -1;
+                    for (const auto& entry : labelCount) {
+                        if (entry.second > maxCount) {
+                            maxCount = entry.second;
+                            bestLabel = entry.first;
+                        }
+                    }
+
+                    // Assign the most frequent label to the test point
+                    testPredictions[i] = bestLabel;
+                }
+            }
+        }
+
+    
+}
+
+vector<vector<int>> readCSV(const string& filename) {
+    vector<vector<int>> data;
+    ifstream file(filename);
+    string line;
+    
+    if (!file.is_open()) {
+        throw runtime_error("Could not open file");
+    }
+
+
+    while (getline(file, line)) {
+        vector<int> row;
+        stringstream ss(line);
+        string value;
+        
+        while (getline(ss, value, ',')) {
+            try {
+                int intValue = stoi(value);
+                row.push_back(intValue);
+            } catch (const invalid_argument& e) {
+                cerr << "Invalid argument: " << e.what() << " for value " << value << '\n';
+            } catch (const out_of_range& e) {
+                cerr << "Out of range: " << e.what() << " for value " << value << '\n';
+            }
+        }
+        data.push_back(row);
+    }
+
+    file.close();
+    return data;
+}
+
+void printData(int **dataArray, int row, int col)
+{
+    for (int i = 0; i < row; i++)
+    {
+        for (int j = 0; j < col; j++)
+        {
+            cout << dataArray[i][j] << "\t";
+        }
+        cout << "\n";
+    }
+}
+
+/**
+ * @brief Main of the Host Application.
+ */
+int main(int argc, char **argv)
+{
+    struct Params params = input_params(argc, argv);
+
+
+    if (params.inputTestFile == nullptr)
+    {
+        //getMatrix(params.numTestPoints, params.dimension, 0, testPoints);
+        getMatrix(params.numTestPoints, params.dimension, testPoints);
```

Comment:
```suggestion
        getMatrix<int>(params.numTestPoints, params.dimension, testPoints);
```
Please note that, these functions are declared as templates, so you need to add type whenever you call these functions.